functionary/vllm_monkey_patch/async_llm_engine.py

```diff
@@ -0,0 +1,724 @@
+
+    async def wait_for_new_requests(self):
+        await self.new_requests_event.wait()
+
+
+class _AsyncLLMEngine(LLMEngine):
+    """Extension of LLMEngine to add async methods."""
+
+    # This is a dict mapping request_id to the list of tools/functions details
+    tools_or_functions: dict = {}
+    # This is a dict mapping request_id to the prompt_template_cls
+    prompt_templates: dict = {}
+    # This is a dict mappingg request_id to the generation_state. It contains
+    # the following information:
+    # - stage: one of the following:
+    # ["pre-function", "function", "pre-parameter", "parameter-name", "parameter-value"]
+    # - curr_tokens: all the tokens for the current function call being generated
+    # - curr_text: curr_tokens but in string text form
+    # - func_name: the function name, if any
+    # - param_names: the parameters names, if any
+    gen_states: dict = {}
+
+    def initialize_gen_state(self, request_id):
+        """This function initializes a new generation state if the request_id
+        is not found in self.gen_states
+        """
+        # Prompt template v2 directly begins with "function" stage while
+        # Prompt template v1 needs to generate <|START_OF_FUNCTION_CALL|> first
+        # Future versions are assumed to begin directly with "function" stage
+        self.gen_states[request_id] = {
+            "stage": "function"
+            if self.prompt_templates[request_id].version not in ["v1"]
+            else "pre-function",
+            "curr_tokens": [],
+            "curr_text": "",
+            "func_name": "",
+            "param_names": [],
+        }
+
+    def update_gen_state(self, request_id, new_token_id):
+        """This function updates the generation state by calling the respective
+        prompt_template.update_grammar_sampling_gen_state class. Each prompt
+        template has its own way of parsing the generation to figure out the
+        current state of generation.
+        """
+        gen_state = self.gen_states[request_id]
+        prompt_template = self.prompt_templates[request_id]
+
+        # Form the functions/parameters options
+        if gen_state["stage"] in ["pre-function", "function"]:
+            options = [
+                tool_or_func["name"]
+                for tool_or_func in self.tools_or_functions[request_id]
+            ]
+        else:
+            func_name = gen_state["func_name"]
+            for tool_or_func in self.tools_or_functions[request_id]:
+                if tool_or_func["name"] == func_name:
+                    options = list(tool_or_func["parameters"]["properties"].keys())
+                    break
+
+        # Update the gen_state
+        self.gen_states[request_id] = prompt_template.update_grammar_sampling_gen_state(
+            gen_state=gen_state,
+            new_token_id=new_token_id,
+            options=options,
+            tokenizer=self.tokenizer,
+        )
+
+    def sample(
```

Comment:
Can we implement this in prompt_template ? So in the future, we can `re-use `this for other framework such as: llama_cpp, HF instead of binding to vllm 
def sample_gramma_token(self, state, delta_token_ids, model_sampled_token_id)

Reply:
Gotcha